### PR TITLE
Reduce number of files opened during testing

### LINKED ISSUE #1950: Testing breaks due to too many open files
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:**  `develop @ a5bbe3a`

## :package: Packages

<details>
<summary>List of packages</summary>

```none
acme==2.6.0
annotated-types==0.6.0
appdirs==1.4.4
Arpeggio==2.0.2
attrs==23.2.0
azure-common==1.1.28
azure-core==1.30.1
azure-identity==1.16.1
azure-keyvault-certificates==4.8.0
azure-keyvault-keys==4.9.0
azure-keyvault-secrets==4.8.0
azure-mgmt-automation==1.0.0
azure-mgmt-compute==30.6.0
azure-mgmt-containerinstance==10.1.0
azure-mgmt-core==1.4.0
azure-mgmt-dns==8.1.0
azure-mgmt-keyvault==10.3.0
azure-mgmt-msi==7.0.0
azure-mgmt-network==25.3.0
azure-mgmt-rdbms==10.1.0
azure-mgmt-resource==23.0.1
azure-mgmt-storage==21.1.0
azure-storage-blob==12.19.1
azure-storage-file-datalake==12.14.0
azure-storage-file-share==12.15.0
certifi==2024.2.2
cffi==1.16.0
charset-normalizer==3.3.2
chevron==0.14.0
click==8.1.7
cryptography==42.0.6
-e git+ssh://git@github.com/jemrobinson/data-safe-haven.git@a5bbe3aaea2b20f8e44d854667943b35dcfc0ef2#egg=data_safe_haven
dill==0.3.8
dnspython==2.4.2
fqdn==1.5.1
freezegun==1.5.1
grpcio==1.60.1
idna==3.7
iniconfig==2.0.0
isodate==0.6.1
josepy==1.14.0
markdown-it-py==3.0.0
mdurl==0.1.2
msal==1.28.0
msal-extensions==1.1.0
msrest==0.7.1
oauthlib==3.2.2
packaging==24.0
parver==0.5
pluggy==1.5.0
portalocker==2.8.2
protobuf==4.25.3
psycopg==3.1.18
pulumi==3.115.0
pulumi_azure_native==2.39.0
pulumi_random==4.16.1
pycparser==2.22
pydantic==2.7.1
pydantic_core==2.18.2
Pygments==2.18.0
PyJWT==2.8.0
pyOpenSSL==24.1.0
pyRFC3339==1.1
pytest==8.2.0
pytest-asyncio==0.23.6
python-dateutil==2.9.0.post0
pytz==2023.4
PyYAML==6.0.1
requests==2.31.0
requests-oauthlib==2.0.0
rich==13.7.1
semver==2.13.0
setuptools==69.5.1
shellingham==1.5.4
simple-acme-dns==3.0.0
six==1.16.0
typer==0.12.3
typing_extensions==4.11.0
urllib3==2.2.1
validators==0.22.0
websocket-client==1.8.0
```

</details>

## :no_entry_sign: Describe the problem

With the recent change in defaults to write everything to a log file, running the tests opens one log file for each test, which eventually causes Python to crash.

Weirdly, this isn't a problem in CI - possibly due to different Python configuration than my local machine?

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
OSError: [Errno 24] Too many open files: '/private/var/folders/dg/n8x2zdyd7pb74pg355wx867c0000gq/T/pytest-of-jrobinson/pytest-190'
```

</details>

## :recycle: To reproduce

Run `hatch run test:test`

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Avoid opening a new log file for each test

### :closed_umbrella: Related issues

Closes #1950 

### :microscope: Tests

Tested locally and in CI.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def local_pulumi_login():
     run([pulumi_path, "logout"], check=False)
 
 
-@fixture(autouse=True)
+@fixture()
```
Comment (JimMadge) [inline tests/conftest.py]: I think we want autouse here, otherwise tests will log to your 'real' log file.

**Review by JimMadge — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 1 above)
Comment (jemrobinson) [reply tests/conftest.py]: This is the thing that causes tests to fail! Every test opens a new log file and doesn't close it until pytest is finished. This means that once you get to N tests, pytest fails due to too many open filehandles.

Perhaps we want a single common log file for tests? I can look into that.

Comment (JimMadge) [reply tests/conftest.py]: Hopefully f17fbec works

Comment (JimMadge) [reply tests/conftest.py]: No 🙃

File: tests/conftest.py [hunk 2]
```diff
@@ -37,18 +37,17 @@ def local_pulumi_login():
...
+def log_directory(session_mocker, tmp_path_factory):
     session_mocker.patch.object(
         data_safe_haven.logging.logger, "logfile_name", return_value="test.log"
     )
+    log_dir = tmp_path_factory.mktemp("logs")
+    session_mocker.patch.object(
+        data_safe_haven.directories, "log_dir", return_value=log_dir
+    )
+    init_logging()
+    return log_dir
```
Comment (jemrobinson) [inline tests/conftest.py]: :D I just came up with the same potential solution. Does it work?

Comment (jemrobinson) [reply tests/conftest.py]: No - this function isn't getting called by any tests that don't explicitly call it.

Comment (JimMadge) [reply tests/conftest.py]: I think so 🤞 :smile:

**Review by JimMadge — APPROVED** [review]

**Pull request merged**